src/h3lib/lib/h3Index.c

```diff
@@ -779,6 +779,86 @@ void H3_EXPORT(h3ToGeoBoundary)(H3Index h3, GeoBoundary* gb) {
                           H3_EXPORT(h3IsPentagon)(h3), gb);
 }
 
+/**
+ * Returns the max number of possible icosahedron faces an H3 index
+ * may intersect.
+ *
+ * @return int count of faces
+ */
+int H3_EXPORT(maxFaceCount)(H3Index h3) {
+    // a pentagon always intersects 5 faces, a hexagon never intersects more
+    // than 2 (but may only intersect 1)
+    return H3_EXPORT(h3IsPentagon)(h3) ? 5 : 2;
+}
+
+/**
+ * Find all icosahedron faces intersected by a given  H3 index, represented
+ * as integers from 0-19. The array is sparse; since 0 is a valid value,
+ * invalid array values are represented as -1. It is the responsibility of
+ * the caller to filter out invalid values.
+ *
+ * @param h3 The H3 index
+ * @param out Output array. Must be of size maxFaceCount(h3).
+ */
+void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+    int res = H3_GET_RESOLUTION(h3);
+    int isPentagon = H3_EXPORT(h3IsPentagon)(h3);
+
+    // We can't use the vertex-based approach here for class II pentagons,
+    // because all their vertices are on the icosahedron edges. Their
+    // direct child pentagons cross the same faces, so use those instead.
+    if (isPentagon && !isResClassIII(res)) {
```

Comment:
Res 15 is a Class III grid as I recall, but can we add one test for a Res 15 pentagon to confirm that it doesn't accidentally try to get a child below Res 15 (which might break `makeDirectChild` is interesting ways)?

Reply:
I did think of this, but you're right, I'll add a test.

Reply:
Test added.

Reply:
I don't think makeDirectChild is well defined for resolution 15 indexes. This call is overflowing the resolution bitfield, and creates a hexagon index at resolution 0 (so with invalid indexing digits) with an "edge/reserved" field of 1. I suspect what happens here is that it needs to recurse again to get to resolution 1, but didn't test. I don't think that code flow is intended, so it may be better to express this in a different way.

Reply:
> I don't think makeDirectChild is well defined for resolution 15 indexes. This call is overflowing the resolution bitfield

This is inside a `!isResClassIII(res)` condition; res 15 is Class III, so this will never be invoked for res 15 hexes. I can add a comment, just in case the condition is changed later.

Reply:
Comment added.

Reply:
I might have misread when I mentioned about resolution 15. I agree this doesn't need a change since it won't be invoked for resolution 15.